go/src/url_shortener/url_shortener.go

```diff
@@ -67,38 +68,46 @@ func (c *URLShortener) incrementHandlerCounter(handler string, succeeded bool) {
 	c.mux.Lock()
 	defer c.mux.Unlock()
 
-	statistics := c.statistics
-
-	statistics.handlerCalls[handler]++
+	c.statistics.handlerCalls[handler]++
 
 	if succeeded {
-		statistics.succeededRedirects++
+		c.statistics.succeededRedirects++
 	} else {
-		statistics.failedRedirects++
+		c.statistics.failedRedirects++
 	}
 }
 
-func (c *URLShortener) getStatistics() string {
-	c.mux.Lock()
-	defer c.mux.Unlock()
+func (s statsJSON) String() string {
+	stats := s.ServerStats
 
-	statistics := c.statistics
-	pairsInCache := fmt.Sprintf("Number of long/short URL pairs: %v", len(c.mappings))
-	succeededRedirects := fmt.Sprintf("Succeeded redirects: %v", statistics.succeededRedirects)
-	failedRedirects := fmt.Sprintf("Failed redirects: %v", statistics.failedRedirects)
+	pairsInCache := fmt.Sprintf("Number of long/short URL pairs: %v", stats.TotalURL)
+	succeededRedirects := fmt.Sprintf("Succeeded redirects: %v", stats.Redirects.Success)
+	failedRedirects := fmt.Sprintf("Failed redirects: %v", stats.Redirects.Failed)
 
-	handlerCalls := make([]string, 0, len(statistics.handlerCalls))
+	handlerCalls := make([]string, 0, len(stats.Handlers))
 
-	for handlerURL, counter := range statistics.handlerCalls {
-		handlerCalls = append(handlerCalls, fmt.Sprintf("Handler %s called %v time(s)", handlerURL, counter))
+	for _, handler := range stats.Handlers {
+		handlerCalls = append(handlerCalls, fmt.Sprintf("Handler %s called %v time(s)", handler.Name, handler.Count))
 	}
 
-	statisticsOut := fmt.Sprintf("Some statistics:\n\n%s\n%s\n%s\n%s", pairsInCache, succeededRedirects, failedRedirects, strings.Join(handlerCalls, "\n"))
-
-	return statisticsOut
+	statsBody := fmt.Sprintf("Some statistics:\n\n%s\n%s\n%s\n%s", pairsInCache, succeededRedirects, failedRedirects, strings.Join(handlerCalls, "\n"))
+	return statsBody
 }
 
-// handlers
+func (c *URLShortener) computeStatistics(stats *statsJSON) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
```

Comment:
See other comment about adding methods to the `urlStatistics` struct.

We are locking the whole server _just_ to compute statistics. However if `urlStatistics`  and `URLshortener` were bit more decoupled we wouldn't have to do that, we'd only have to lock the stats struct to update the statistics